### ISSUE TITLE
First class support for realworld bidirectional messaging

### ISSUE THREAD

**Issue by prabirshrestha** [salvo-rs/salvo#185]
Salvo already supports websockets but would it be possible to provide a support for better bidirectional messaging via some sort of well known api/protocol.

Rather than creating your own protocol, would be good to support existing ones

*  [Phoenix Channels](https://hexdocs.pm/phoenix/channels.html). Trillium's implementation can be found at https://github.com/trillium-rs/trillium/tree/main/channels.
* [Rust socket.io](https://github.com/1c3t3a/rust-socketio)
* [SockJs](https://github.com/sockjs/sockjs-client#sockjs-family). This is probably the simpler one out there so should be easier to implement as well as it has log of implementation in various languages. Actix implementation can be found [here](https://github.com/actix/sockjs).

Advanatages over pure websockets:
* xhrpolling/xh-streaming fallback
* Easily swap the implementation to redis/or some other cloud specific internal messaging queues.
* Auth support without writing custom one.

This allows me to focus on the business logic instead of internal details.

**Comment by prabirshrestha** [comment]
Might be collaborate with this project [socketioxide](https://github.com/Totodore/socketioxide) as [socket.io](https://socket.io/) is a popular protocol. It seems it already supports Axum/Warp and Hyper. Would be good to have it working Salvo as first-class citizen. 

//cc @Totodore

**Comment by Totodore** [comment]
I'll be glad to add salvo to the list of the supported framework for socketioxide.

The only thing I need is to add a hyper service wrapper or a tower layer to the router. It seems a bit different than Axum or Warp so maybe some glue code will be required.

Also note that socketioxide is not production ready and that it is under heavy development.

**Comment by chrislearn** [comment]
> I'll be glad to add salvo to the list of the supported framework for socketioxide.
> 
> The only thing I need is to add a hyper service wrapper or a tower layer to the router. It seems a bit different than Axum or Warp so maybe some glue code will be required.
> 
> Also note that socketioxide is not production ready and that it is under heavy development.

It does look a bit difficult, let me think for a while。

**Comment by chrislearn** [comment]
@Totodore @prabirshrestha 

I implemented a support compatible with Tower service and layer.
However, salvo is now based on hyper 1.0-rc4, which will cause many middleware based on hyper 0.14.x tower to be incompatible, which should also include socketioxide.

**Comment by Totodore** [comment]
That's really nice ! It will also allows anyone to add custom tower layers to salvo ! 

You're right, socketioxide is currently based on hyper 0.14.

I'll add a tracking issue to prepare a working branch on 1.0-rc*.

**Comment by Totodore** [comment]
#468 is blocking me at the moment

**Comment by Totodore** [comment]
The integration of hyper v1.0 is done for socketioxide. I have added an [example with salvo](https://github.com/Totodore/socketioxide/blob/main/examples/salvo-echo/salvo_echo.rs) on the main branch.

I plan to release a new version which will include the hyper 1.0 support in the next week. I want to improve examples and documentation before.

**Comment by Totodore** [comment]
> The integration of hyper v1.0 is done for socketioxide. I have added an [example with salvo](https://github.com/Totodore/socketioxide/blob/main/examples/salvo-echo/salvo_echo.rs) on the main branch.
> 
> I plan to release a new version which will include the hyper 1.0 support in the next week. I want to improve examples and documentation before.

I just release socketioxide v0.7.0 with supports for hyper v1. It would be really nice to add the example in your repo too :). I can make a PR if you it's ok for you.

**Comment by prabirshrestha** [comment]
Tested this and everything seems to be working.

I do see the error log from salvo when integrating with and existing project.

```
2023-11-12T21:24:47.913550Z ERROR salvo_core::tower_compat: `FlowCtrlOutContext` should exists in response extensions.
```

**Comment by prabirshrestha** [comment]
any idea on why I would get the above error message?

**Comment by chrislearn** [comment]
> any idea on why I would get the above error message?

Fixed via #503

**Comment by chrislearn** [comment]
@Totodore salvo 0.63 released base on hyper 1.0.
